plugins/inputs/t128_graphql/t128_graphql_test.go

```diff
@@ -0,0 +1,442 @@
+	InvalidFieldQuery             = "query MyQuery{allRouters(name:\"ComboEast\"){nodes{nodes(name:\"east-combo\"){nodes{arp{nodes{\ninvalid-field\n}}}}}}}"
+)
+
+var ResponseProcessingTestCases = []struct {
+	Name            string
+	EntryPoint      string
+	Fields          map[string]string
+	Tags            map[string]string
+	Query           string
+	Endpoint        Endpoint
+	ExpectedMetrics []*testutil.Metric
+	ExpectedErrors  []string
+}{
+	{
+		Name:       "empty query produces error",
+		EntryPoint: "",
+		Fields:     nil,
+		Tags:       nil,
+		Query:      "query MyQuery{{\n}}",
+		Endpoint: Endpoint{"/api/v1/graphql", 400, `{"query":"query MyQuery{{\n}}"}`, `
+		{
+			"errors": [{
+				"name": "BadRequestError",
+				"message": "Must provide query string."
+			}]
+		}
+		`},
+		ExpectedMetrics: nil,
+		ExpectedErrors: []string{
+			"status code 400 not OK for metric test-metric: Must provide query string.",
+		},
+	},
+	{
+		Name:            "empty response produces error",
+		EntryPoint:      "allRouters[name:ComboEast]/nodes/nodes[name:east-combo]/nodes/arp/nodes",
+		Fields:          map[string]string{"test-field": "test-field"},
+		Tags:            map[string]string{},
+		Query:           ValidQueryNoTag,
+		Endpoint:        Endpoint{"/api/v1/graphql", 200, ValidExpectedRequestNoTag, "{}"},
+		ExpectedMetrics: nil,
+		ExpectedErrors: []string{
+			"unexpected response for metric test-metric: {}",
+		},
+	},
+	{
+		Name:       "none value produces error",
+		EntryPoint: "allRouters[name:ComboEast]/nodes/nodes[name:east-combo]/nodes/arp/nodes",
+		Fields:     map[string]string{"test-field": "test-field"},
+		Tags:       map[string]string{},
+		Query:      ValidQueryNoTag,
+		Endpoint: Endpoint{"/api/v1/graphql", 200, ValidExpectedRequestNoTag, `{
+			"data": {
+				"allRouters": {
+				  	"nodes": [{
+					  	"nodes": {
+							"nodes": [{
+								"arp": {
+							  		"nodes": [{
+								  		"test-field": null
+									}]
+								}
+						  	}]
+					  	}
+					}]
+				}
+			}
+		}`},
+		ExpectedMetrics: nil,
+		ExpectedErrors: []string{
+			"found empty data for metric test-metric: field test-field",
+		},
+	},
+	{
+		Name:       "converts tag to string if numeric",
+		EntryPoint: "allRouters[name:ComboEast]/nodes/nodes[name:east-combo]/nodes/arp/nodes",
+		Fields:     map[string]string{"test-field": "test-field"},
+		Tags:       map[string]string{"test-tag": "test-tag"},
+		Query:      ValidQuerySingleTag,
+		Endpoint: Endpoint{"/api/v1/graphql", 200, ValidExpectedRequestSingleTag, `{
+			"data": {
+				"allRouters": {
+				  	"nodes": [{
+					  	"nodes": {
+							"nodes": [{
+								"arp": {
+							  		"nodes": [{
+								  		"test-field": 128,
+								  		"test-tag": 128
+									}]
+								}
+						  	}]
+					  	}
+					}]
+				}
+			}
+		}`},
+		ExpectedMetrics: []*testutil.Metric{
+			{
+				Measurement: "test-metric",
+				Tags:        map[string]string{"test-tag": "128"},
```

Comment:
@gregschrock tags can't be ints?

Reply:
I'm not Greg, but yes, tags can only be strings